ADMIN-PORT.md

```diff
@@ -0,0 +1,46 @@
+---
+RFC: 3
+Status: Proposed
+---
+
+# ADMIN PORT RFC
+
+## Abstract
+
+In Valkey.conf, Add a new config parameter: "admin-port" to allow cloud providers or internal administrators in a company or special users from public internet to connect to the Valkey server to execute some special commands or load specfic modules.
+
+
+## Motivation
+
+For adminstrators, they may want to execute some special commands in the server, which are usually defined in the module, and
+other general clients are not allowed to run these kinds of commands. Thus connecting to an admin-port is a good way to implement this feature. 
+
+Beside above mentioned, we could implement the following features in the future once admin-port is commited:
+1. The connections through admin port are not limited to maxclient number, we can always guarantee the administrators can connect to the server.
+2. Some status check commands could be sent and receied through admin port, we can do some special logic to make sure they have opportunity to be handle first, such as: PING, HELLO, ECHO, INFO etc
```

Comment:
Are you suggesting we add some code like this?

```C
if (c->cmd->proc == pingCommand || 
    c->cmd->proc == helloCommand ||
    c->cmd->proc == echoCommand ||
    c->cmd->proc == infoCommand) {
    /* Permission OK */
}
```

I don't like a hard-coded list of commands like this. It is better that it matched to one ACL category or something like that. Or maybe configurable.

Reply:
Then I think the RFC should describe how the module can check this. We need a new module API? Or a command flag to ValkeyModule_CreateCommand?

Reply:
In the Specification part, I add several words: You can find a way to to get the port number that client connects to the server and then to decide if the command can be called by the client.

In fact, developers have a lot of ways to get the client connected port, and in our side, we have one way to check the client connected node ipaddress and port. And I believe there are other ways to implement the same function. Thus, in this RFC, I just want to indicate the port check should happen in the function which is defined in RedisModule_CreateCommand

Reply:
I don't know how a module get this information. I think it's good to describe some example how it can be done in the design considerations. It is useful context to understand the feature.

You have to call `ValkeyModule_Call("CLIENT INFO")`?

I would hope the module can find out with `ValkeyModule_GetClientInfoById` but it seems to return only the peer port and IP, not the server port it connected to.

<details>
<summary>ValkeyModule_GetClientInfoById</summary>

```
              int ValkeyModule_GetClientInfoById(void *ci, uint64_t id);

       Available since: 6.0.0

       Return information about the client with the  specified  ID  (that  was
       previously  obtained  via  the ValkeyModule_GetClientId() API).  If the
       client exists, VALKEYMODULE_OK is returned, otherwise  VALKEYMODULE_ERR
       is returned.

       When  the  client exist and the ci pointer is not NULL, but points to a
       structure of type ValkeyModuleClientInfoV1, previously initialized with
       the correct VALKEYMODULE_CLIENTINFO_INITIALIZER_V1,  the  structure  is
       populated with the following fields:

               uint64_t flags;         // VALKEYMODULE_CLIENTINFO_FLAG_*
               uint64_t id;            // Client ID
               char addr[46];          // IPv4 or IPv6 address.
               uint16_t port;          // TCP port.
               uint16_t db;            // Selected DB.

       Note:  the  client  ID is useless in the context of this call, since we
       already know, however the same structure could be used  in  other  con‐
       texts  where we don’t know the client ID, yet the same structure is re‐
       turned.

       With flags having the following meaning:

              VALKEYMODULE_CLIENTINFO_FLAG_SSL          Client using SSL connection.
              VALKEYMODULE_CLIENTINFO_FLAG_PUBSUB       Client in Pub/Sub mode.
              VALKEYMODULE_CLIENTINFO_FLAG_BLOCKED      Client blocked in command.
              VALKEYMODULE_CLIENTINFO_FLAG_TRACKING     Client with keys tracking on.
              VALKEYMODULE_CLIENTINFO_FLAG_UNIXSOCKET   Client using unix domain socket.
              VALKEYMODULE_CLIENTINFO_FLAG_MULTI        Client in MULTI state.

       However passing NULL is a way to just check if  the  client  exists  in
       case we are not interested in any additional information.

       This  is  the  correct usage when we want the client info structure re‐
       turned:

               ValkeyModuleClientInfo ci = VALKEYMODULE_CLIENTINFO_INITIALIZER;
               int retval = ValkeyModule_GetClientInfoById(&ci,client_id);
               if (retval == VALKEYMODULE_OK) {
                   printf("Address: %s\n", ci.addr);
               }
```
</detail>

Reply:
I just add a block: Sample for Module, which indicates one way to get the admin port and how to use it in module. But I think more way to get it. Please take a look